plasmapy/particles/particle_class.py

```diff
@@ -96,6 +97,59 @@ def _category_errmsg(particle, category: str) -> str:
     return errmsg
 
 
+class ParticleJSONDecoder(json.JSONDecoder):
+    def __init__(self, *args, **kwargs):
+        json.JSONDecoder.__init__(self, object_hook=self.object_hook, *args, **kwargs)
+
+    def object_hook(self, json_dict):
+        """
+        Returns the appropriate particle object corresponding to the input JSON
+        string. If not a particle, returns a dictionary object. Logic Implementation for
+        plasmapy.particles.from_json method.
+
+        Examples
+        --------
+        >>> import plasmapy
+        >>> input_str = '{"plasmapy_particle": {"type": "Particle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"argument": "Pb"}}}'
+        >>> plasmapy.particles.from_json(input_str)
+        Particle("Pb")
+        >>> input_str = '{"plasmapy_particle": {"type": "CustomParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.12 kg", "charge": "6.2 C"}}}'
+        >>> plasmapy.particles.from_json(input_str)
+        CustomParticle(mass=5.12 kg, charge=6.2 C)
+        >>> input_str = '{"plasmapy_particle": {"type": "DimensionlessParticle", "module": "plasmapy.particles.particle_class", "date_created": "2020-06-08", "init_args": {"mass": "5.2", "charge": "6.3"}}}'
+        >>> plasmapy.particles.from_json(input_str)
+        DimensionlessParticle(mass=5.2, charge=6.3)
+        """
+        particle_types = {
+            "CustomParticle": CustomParticle,
+            "DimensionlessParticle": DimensionlessParticle,
+            "Particle": Particle,
+        }
+        if "plasmapy_particle" in json_dict:
+            particle_data = json_dict["plasmapy_particle"]
+            particle_class = particle_data["type"]
+            if particle_class in particle_types:
+                particle_init_args = particle_data["init_args"]
+                if particle_class == "CustomParticle":
+                    particle_init_args["mass"] = u.Quantity(particle_init_args["mass"])
+                    particle_init_args["charge"] = u.Quantity(
+                        particle_init_args["charge"]
+                    )
+                if particle_class == "DimensionlessParticle":
+                    particle_init_args["mass"] = np.float64(particle_init_args["mass"])
+                    particle_init_args["charge"] = np.float64(
+                        particle_init_args["charge"]
+                    )
+                return particle_types[particle_class](**particle_init_args)
+            else:
+                raise InvalidParticleError(
+                    f"Particle with type {particle_class} is not "
+                    f"present in the _particle_types dictionary or is undefined."
+                )
```

Comment:
```suggestion
            try:
                pardict = json_dict["plasmapy_particle"]
                partype = pardict["type"]
                args = pardict["__init__"]["args"]
                kwargs = pardict["__init__"]["kwargs"]
                particle = particle_types[partype](*args, **kwargs)
            except KeyError:
                raise InvalidElementError(
                    f"json file does not define a valid plasmapy particle"
                )
            return particle
```

With the updated `json_dict`, this can run much more succinctly.  However, in-order for this to work with `CustomParticle` the parameter validations need to be updated so they can handle string representations of astropy `Quantity`'s.  I've already figured this out and will open a separate PR on it.  Basically we want the following to work...

```python
par = CustomeParticle(mass="5.2 kg")
```

Reply:
Thank you for suggesting this! It looks much more elegant now.